src/audio/pcm_converter/pcm_converter_hifi3.c

```diff
@@ -757,7 +807,523 @@ const struct pcm_func_map pcm_func_map[] = {
+	ae_int24x2 *in = audio_stream_wrap(source, src + ioffset * 3);
+	ae_int32x2 *out = audio_stream_wrap(sink, dst + ooffset);
+
+	for (left_samples = samples; left_samples; left_samples -= n) {
+		nmax = audio_stream_bytes_without_wrap(source, in) / 3;
+		n = MIN(left_samples, nmax);
+		nmax = audio_stream_samples_without_wrap_s32(sink, out);
+		n = MIN(n, nmax);
+		m = n >> 1;
+		inu = AE_LA64_PP(in);
+		for (i = 0; i < m; i++) {
+			/* load 2 32 bit samples */
+			AE_LA24X2_IP(sample24, inu, in);
+			sample = AE_MOVINT32_FROMINT24X2(sample24);
+			/* shift left and store two 32 bit samples */
+			AE_SA32X2_IP(sample, outu, out);
+		}
+		AE_SA64POS_FP(outu, out);
+
+		/* process the left 1 sample to avoid memory access overrun */
+		if (n & 0x01) {
+			AE_LA24_IP(sample24, inu, in);
+			sample = AE_MOVINT32_FROMINT24X2(sample24);
+			AE_S32_L_IP(sample, (ae_int32 *)out,
+				    sizeof(ae_int32));
+		}
+
+		in = audio_stream_wrap(source, in);
+		out = audio_stream_wrap(sink, out);
+	}
+
+	return samples;
+}
+
+static int pcm_convert_s24_c32_to_s24_c24(const struct audio_stream __sparse_cache *source,
+					  uint32_t ioffset,
+					  struct audio_stream __sparse_cache *sink,
+					  uint32_t ooffset, uint32_t samples)
+{
+	int32_t *src = source->r_ptr;
+	uint8_t *dst = sink->w_ptr;
+	ae_int32x2 sample = AE_ZERO32();
+	ae_int24x2 sample24 =  AE_ZERO24();
+	uint32_t nmax, i, n, m, left_samples;
+	ae_valign outu = AE_ZALIGN64();
+	ae_valign inu = AE_ZALIGN64();
+
+	ae_int32x2 *in = audio_stream_wrap(source, src + ioffset);
+	ae_int24x2 *out = audio_stream_wrap(sink, dst + ooffset * 3);
+
+	for (left_samples = samples; left_samples; left_samples -= n) {
+		nmax = audio_stream_samples_without_wrap_s32(source, in);
+		n = MIN(left_samples, nmax);
+		nmax = audio_stream_bytes_without_wrap(sink, out) / 3;
+		n = MIN(n, nmax);
+		m = n >> 1;
+		inu = AE_LA64_PP(in);
+		for (i = 0; i < m; i++) {
+			/* load 2 32 bit samples */
+			AE_LA32X2_IP(sample, inu, in);
+			sample24 = AE_MOVINT24X2_FROMF32X2(sample);
+			AE_SA24X2_IP(sample24, outu, out);
+		}
+		AE_SA64POS_FP(outu, out);
+
+		/* process the left 1 sample to avoid memory access overrun */
+		if (n & 0x01) {
+			AE_L32_IP(sample, (ae_int32 *)in, sizeof(ae_int32));
+			sample24 = AE_MOVINT24X2_FROMF32X2(sample);
+			AE_SA24_IP(sample24, outu, out);
+		}
+
+		in = audio_stream_wrap(source, in);
+		out = audio_stream_wrap(sink, out);
+	}
+
+	return samples;
+}
+
+/* 2x24bit samples are packed into 3x16bit samples for hda link dma */
+/* Keep C version now since there are some bugs need to fix, waitting for updating from Rander*/
+static int pcm_convert_s24_c32_to_s24_c24_link_gtw(const struct audio_stream __sparse_cache *source,
+						   uint32_t ioffset,
+						   struct audio_stream __sparse_cache *sink,
+						   uint32_t ooffset, uint32_t samples)
+{
+	int32_t *src = source->r_ptr;
+	uint16_t *dst = sink->w_ptr;
+	int processed;
+	int nmax, i = 0, n = 0;
+
+	src += ioffset;
+	assert(ooffset == 0);
```

Comment:
should it really be an assertion or would returning an error make more sense?

Reply:
sorry for not paying much attention in this function since it needs to be correctted. I am not sure about the situation of this limitation, and actually I think it should be dst += ooffset. hi @RanderWang , can you help to explain this?

Reply:
please dont use assert(), it's unrecoverable. log() and error and return an error value up the stack.